[master] Stop loading utils modules into `__utils__` if they don't define `__virtual__`

s0undt3ch opened the pull request:
### What does this PR do?
See title

### What issues does this PR fix or reference?
Refs: #62191

s0undt3ch committed:
Stop using `__utils__` in `salt/grains/zfs.py`

Refs #62191

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Stop loading utils modules into `__utils__` if they don't define `__virtual__`

Refs #62191

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Remove `__utils__` usage from `salt/{modules,states}/{zfs,zpool}.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Remove `__utils__` usage from `salt/modules/network.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Remove `__utils__` usage from `salt/roster/*.py`

Additionally, remove `__opts__` usage in `salt/utils/roster_matcher.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_environ.py` with regards to the `__utils__` changes

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_proxy.py` with regards to the `__utils__` changes

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_win_snmp.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Stop using `__utils__` in `salt/modules/state.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

twangboy commented on salt/roster/terraform.py:
```diff
@@ -253,4 +254,4 @@ def targets(tgt, tgt_type="glob", **kwargs):  # pylint: disable=W0613
     raw = _parse_state_file(roster_file)
     log.debug("%s hosts in terraform state file", len(raw))
-    return __utils__["roster_matcher.targets"](raw, tgt, tgt_type, "ipv4")
+    return salt.utils.roster_matcher.targets(__opts__, raw, tgt, tgt_type, "ipv4")
```
Did you mean to add `__opts__` here?

s0undt3ch replied:
Yes. The utils module _used_ to use `__opts__`.
Now it doesn't, we have to pass them in.

twangboy commented on salt/utils/roster_matcher.py:
```diff
@@ -22,8 +22,8 @@
-def targets(conditioned_raw, tgt, tgt_type, ipv="ipv4"):
-    rmatcher = RosterMatcher(conditioned_raw, tgt, tgt_type, ipv)
+def targets(opts, conditioned_raw, tgt, tgt_type, ipv="ipv4"):
+    rmatcher = RosterMatcher(opts, conditioned_raw, tgt, tgt_type, ipv)
```
Added `opts` intentional?

s0undt3ch replied:
Yep.

s0undt3ch committed:
Fix some more `__utils__` usage in `salt/modules/{chroot,yaml}.py` and `salt/states/{btrfs,file}.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_etcd_mod.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/cloud/clouds/test_scaleway.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/unit/modules/test_nxos.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/unit/utils/test_roster_matcher.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_chroot.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_google_chat.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_s3.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_transactional_update.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/dockermod/test_trans_tar.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/file/test_file_module.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/pillar/test_s3.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/proxy/test_cimc.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/proxy/test_panos.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/proxy/nxos/test_nxos_nxapi.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/states/test_btrfs.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/states/test_win_dns_client.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/states/file/test_prepend.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/win_lgpo/test_adv_audit_settings.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/unit/modules/test_git.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Idem execution and state modules no longer use `__utils__`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Remove unnecessary `__virtual__` functions from "safe" salt utils modules

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Adjust win lgpo code for proper functional testing

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `__utils__` usage in `salt/modules/{scp_mod,travisci}.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix win lgpo tests

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Bubble up any warnings raised into state results.

Any warnings raised by Python's `warnings` module are now included in state returns under the `warnings` key.
This allows for, as an example, Salt deprecation warnings from execution modules to bubble up into state returns as opposed to only be buried in the logs.

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Add changelog and update release notes

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix tests now that we capture warnings and they're returned in state results

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Memoize the ast check logic

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Remove `__utils__` usage from `salt/modules/dockermod.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Show a deprecation warning on __utils__ access that will go away in 3010

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Ignore a `DeprecationWarning` which is breaking the test assertions

```
DeprecationWarning: the load_module() method is deprecated and slated for removal in Python 3.12; use exec_module() instead
```

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

Ch3LL commented on salt/utils/win_lgpo_auditpol.py:
```diff
@@ -298,8 +288,176 @@ def get_auditpol_dump():
+    return audit_defaults
+
+
+def get_advaudit_defaults(option=None):
```
same here, i'm not following why we are adding these new functions?

s0undt3ch replied:
So we can run tests against the functionality

s0undt3ch replied:
Wait, this one is weird, let me check

s0undt3ch replied:
This was a private utility function in the execution module just moved to the utility module.

s0undt3ch commented on salt/states/test.py:
```diff
@@ -478,17 +478,28 @@ def check_pillar(
     return ret
+
+
+def deprecation_warning_raised(name):
```
Se we can run tests against the functionality

Ch3LL reviewed (changes requested):
We should also add a .deprecated.md file to inform users that __utils__ usage will be removed in 3009.0.

s0undt3ch committed:
Stop using `__utils__` in a few windows related module/states/utils

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Call `salt.utils.args.clean_kwargs` directly instead of through `__utils__`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Stop `__utils__` usage in `salt/modules/file.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix any remaining, safe, `__utils__` usage.

The term safe here is, if the utils module uses a salt dunder, then any
calls using `__utils__` to access that module will no be converted to be
used directly since it needs access to the salt dunders, only accessible
when calling the functions through `__utils__`.

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fix `tests/pytests/functional/states/win_lgpo/test_adv_audit_settings_state.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Account for, and test, both pre and post 3009 behaviors

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

twangboy commented:
Looks like there's a problem with the docs build... and this needs to be rebased.

s0undt3ch committed:
Fix `tests/unit/modules/test_vagrant.py`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Add an AST checker to prevent `__utils__` usage

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch commented:
The branch still exists for someone to pick it up, feel free to push it to the Salt repo.
